tests/login.spec.ts

```diff
@@ -1,30 +1,119 @@
-import { test, expect } from './fixtures'
-
-test.describe('Login Functionality', {
-    tag: ['@login', '@critical']
-}, () => {
-    test('TC_LOGIN_001: Valid Login redirects to dashboard', {
-        tag: ['@smoke']
-    }, async ({ loginPage, dashboardPage, validUser }) => {
+import { test, expect } from "./fixtures";
+
+const NEGATIVE_LOGIN_SCENARIOS = [
+  {
+    testId: "TC_LOGIN_002",
+    description: "Invalid Username",
+    fixture: "invalidUser",
+    expectedError: "Invalid credentials"
+  },
+  {
+    testId: "TC_LOGIN_004",
+    description: "Invalid Password",
+    fixture: "invalidPasswordUser",
+    expectedError: "Invalid credentials"
+  }
+];
+
+const VALIDATION_SCENARIOS = [
+  {
+    testId: "TC_LOGIN_005",
+    description: "Empty Username",
+    fixture: "emptyUsernameUser",
+    expectedMessage: "singleRequiredMessage"
+  },
+  {
+    testId: "TC_LOGIN_006",
+    description: "Empty Password",
+    fixture: "emptyPasswordUser",
+    expectedMessage: "singleRequiredMessage"
+  }
+];
+
+const SECURITY_SCENARIOS = [
+  {
+    testId: "TC_LOGIN_008",
+    description: "SQL Injection Test",
+    fixture: "sqlInjectionUser",
+    expectedError: "Invalid credentials"
+  },
+  {
+    testId: "TC_LOGIN_009",
+    description: "XSS Attack Test",
+    fixture: "xssAttackUser",
+    expectedError: "Invalid credentials"
+  }
+];
+
+test.describe(
+  "Login Functionality",
+  {
+    tag: ["@login", "@critical"]
+  },
+  () => {
+    test(
+      "TC_LOGIN_001: Valid Login redirects to dashboard",
+      {
+        tag: ["@smoke"]
+      },
+      async ({ loginPage, dashboardPage, validUser }) => {
         await loginPage.goto();
         await loginPage.login(validUser.username, validUser.password);
-        
+
         await expect(loginPage.page).toHaveURL(/dashboard/, { timeout: 10000 });
         await expect(dashboardPage.dashboardHeading).toBeVisible();
-    });
-
-    test('TC_LOGIN_002: Invalid Username, "Invalid credentials" is displayed', {
-        tag: ['@negative']
-    }, async ({ loginPage, invalidUser }) => {
-        await loginPage.goto();
-        await loginPage.login(invalidUser.username, invalidUser.password);
-
-        await expect(loginPage.errorMessage).toBeVisible({ timeout: 5000 });
-    });
-
-    test('TC_LOGIN_003: Forgot Password Link', {
-        tag: ['@navigation']
-    }, async ({ loginPage, forgotPasswordPage }) => {
+      }
+    );
+
+    test.describe(
+      "Invalid Credentials Tests",
+      {
+        tag: ["@negative"]
+      },
+      () => {
+        NEGATIVE_LOGIN_SCENARIOS.forEach(
+          ({ testId, description, fixture, expectedError }) => {
+            if (fixture === "invalidUser") {
```

Comment:
`if ... else` is the wrong way to do that, and you don't achieve any benefit compared to the lack of parameterisation